src/settings/toml/deploy_target.rs

```diff
@@ -0,0 +1,155 @@
+use crate::settings::toml::Route;
+
+#[derive(Debug)]
+pub struct RouteConfig {
+    pub workers_dev: Option<bool>,
+    pub route: Option<String>,
+    pub routes: Option<Vec<String>>,
+    pub zone_id: Option<String>,
+    pub account_id: Option<String>,
+}
+
+impl RouteConfig {
+    fn has_conflicting_targets(&self) -> bool {
+        if self.workers_dev.unwrap_or_default() {
+            self.routes_defined()
+        } else {
+            if let Some(pattern) = &self.route {
+                !pattern.is_empty() && self.routes.is_some()
+            } else {
+                false
+            }
+        }
+    }
+
+    pub fn routes_defined(&self) -> bool {
+        if self.routes.is_some() {
+            true
+        } else if let Some(pattern) = &self.route {
+            !pattern.is_empty()
+        } else {
+            false
+        }
+    }
+
+    pub fn is_zoneless(&self) -> bool {
+        self.workers_dev.unwrap_or_default() && !self.has_conflicting_targets()
+    }
+
+    pub fn is_zoned(&self) -> bool {
+        self.routes_defined() || !self.is_missing_zone_id()
+    }
+
+    // zone id is another weird one where `Some("")` is treated the same as `None`
+    pub fn is_missing_zone_id(&self) -> bool {
+        if let Some(zone_id) = &self.zone_id {
+            zone_id.is_empty()
+        } else {
+            true
+        }
+    }
+
+    // account id is another weird one where `Some("")` is treated the same as `None`
+    pub fn is_missing_account_id(&self) -> bool {
+        if let Some(account_id) = &self.account_id {
+            account_id.is_empty()
+        } else {
+            true
+        }
+    }
+
+    pub fn workers_dev_false_by_itself(&self) -> bool {
+        if let Some(workers_dev) = self.workers_dev {
+            !workers_dev && !self.routes_defined()
+        } else {
+            false
+        }
+    }
+}
+
+impl DeployTarget {
+    pub fn build(
+        script_name: &str,
+        route_config: &RouteConfig,
+    ) -> Result<DeployTarget, failure::Error> {
+        if route_config.is_zoneless() {
+            // account_id is required
+            let account_id = route_config.account_id.as_ref().unwrap();
```

Comment:
should probably replace this with an `unwrap_or_else(|| failure::bail!()` if possible or an expect